plugins/hashsum_download/server/__init__.py

```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+###############################################################################
+#  Copyright Kitware Inc.
+#
+#  Licensed under the Apache License, Version 2.0 ( the "License" );
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+###############################################################################
+
+from girder.api import access
+from girder.api.describe import describeRoute, Description
+from girder.api.rest import RestException
+from girder.api.v1.file import File
+
+
+class HashedFile(File):
+
+    SupportedAlgorithms = [
+        'sha512',
+    ]
+
+    def __init__(self, apiRoot):
+        super(File, self).__init__()
+
+        self.resourceName = 'file'
+        apiRoot.file.route('GET', ('hashsum', ':algo', ':hash', 'download'),
+                           self.downloadWithHash)
+
+    @access.public
+    @describeRoute(
+        Description('Download a file by its hashsum.')
+        .param('algo', 'The type of the given hashsum.',
+               paramType='path', enum=SupportedAlgorithms)
+        .param('hash', 'The hashsum of the file to do wnload.',
+               paramType='path')
+        .errorResponse()
+        .errorResponse('Read access was denied on the file.', 403)
+    )
+    def downloadWithHash(self, algo, hash, params):
+        algo = algo.lower()
+        if algo not in self.SupportedAlgorithms:
+            msg = 'Invalid algorithm ("%s"). Supported algorithm are: %s.'\
+                  % (algo, self.SupportedAlgorithms)
+            raise RestException(msg, code=400)
+
+        query = {algo: hash.lower()}  # Always convert to lower case
+        fileModel = self.model('file')
+        file = fileModel.findOne(query)
+        if not file:
+            raise RestException('File not found.', code=404)
```

Comment:
@zachmullen I don't want to sidetrack this too much, but I'm not totally clear on what you're asking for. Here's what we do now for every other route:
1. Unmatched route -> 404
2. Matched route, ObjectId syntax invalid -> 400
3. Matched route, ObjectId syntax valid, document missing from DB -> 400
4. Document found, anonymous user access denied -> 401
5. Document found, logged-in user access denied -> 403
6. Document found, logged-in user access denied -> 200

If your concern is that we are leaking information about the presence of absence of documents to unauthorized users, that's legitimate, but it's a deeper problem with the fact that [``AccessControlledModel.load``](https://github.com/girder/girder/blob/master/girder/models/model_base.py#L915) raises either a ``ValidationException`` on 2 or 3 or an ``AccessException`` (from [``AccessControlledModel.requireAccess``](https://github.com/girder/girder/blob/master/girder/models/model_base.py#L894); also this should be noted in docstring of ``AccessControlledModel.load``) on 4 or 5, then status codes are set in [``_handleRestException``](https://github.com/girder/girder/blob/master/girder/api/rest.py#L366) or [``_handleAccessException``](https://github.com/girder/girder/blob/master/girder/api/rest.py#L375).

So, all I'm proposing is that for this route, an unfound hash value (case 3, since we don't check hash syntax) returns 404. Everything else will be the same as the rest of the API (for better or worse).